File: apps/android/app/src/main/java/com/termirust/mobile/controller/MobileRouteContract.kt
package com.termirust.mobile.controller

enum class MobileItemKind(val wireValue: String) {
    SAVED_CONNECTION("saved_connection"),
    PAIRED_DEVICE("paired_device"),
    DURABLE_DEVICE_SESSION("durable_device_session"),
}

enum class MobileCredentialOwner(val wireValue: String) {
    SSH_CREDENTIAL("ssh_credential"),
    DEVICE_PAIRING_IDENTITY("device_pairing_identity"),
}

enum class MobileContinuityOwner(val wireValue: String) {
    NONE("none"),
    REMOTE_TMUX_IF_ENABLED("remote_tmux_if_enabled"),
    HOST_SERVICE("host_service"),
}

enum class MobileRouteCapability(val wireValue: String) {
    LIST_DEVICE_SESSIONS("list_device_sessions"),
    TERMINAL_OUTPUT("terminal_output"),
    TERMINAL_INPUT("terminal_input"),
    TERMINAL_RESIZE("terminal_resize"),
    PERSISTENT_TMUX("persistent_tmux"),
    DURABLE_REPLAY("durable_replay"),
    AUTHORITATIVE_ACTIVITY("authoritative_activity"),
    SINGLE_WRITER("single_writer"),
}

enum class MobileRouteContractError(val wireValue: String) {
    UNKNOWN_ITEM_KIND("unknown_item_kind"),
    UNKNOWN_CAPABILITY("unknown_capability"),
    CREDENTIAL_OWNER_MISMATCH("credential_owner_mismatch"),
    CONTINUITY_OWNER_MISMATCH("continuity_owner_mismatch"),
    CAPABILITY_MISMATCH("capability_mismatch"),
    TERMINAL_OWNERSHIP_MISMATCH("terminal_ownership_mismatch"),
}

class MobileRouteContractException(val reason: MobileRouteContractError) : IllegalArgumentException(reason.wireValue)

data class MobileRouteProjection(
    val itemKind: MobileItemKind,
    val credentialOwner: MobileCredentialOwner,
    val continuityOwner: MobileContinuityOwner,
    val capabilities: Set<MobileRouteCapability>,
    val canOpenTerminal: Boolean,
) {
    companion object {
        fun validated(
            itemKind: String,
            credentialOwner: String,
            continuityOwner: String,
            capabilities: List<String>,
            canOpenTerminal: Boolean,
        ): MobileRouteProjection {
            val kind = MobileItemKind.entries.firstOrNull { it.wireValue == itemKind }
                ?: fail(MobileRouteContractError.UNKNOWN_ITEM_KIND)
            val credential = MobileCredentialOwner.entries.firstOrNull { it.wireValue == credentialOwner }
                ?: fail(MobileRouteContractError.CREDENTIAL_OWNER_MISMATCH)
            val continuity = MobileContinuityOwner.entries.firstOrNull { it.wireValue == continuityOwner }
                ?: fail(MobileRouteContractError.CONTINUITY_OWNER_MISMATCH)
            val parsed = capabilities.map { raw ->
                MobileRouteCapability.entries.firstOrNull { it.wireValue == raw }
                    ?: fail(MobileRouteContractError.UNKNOWN_CAPABILITY)
            }.toSet()
            if (parsed.size != capabilities.size) fail(MobileRouteContractError.CAPABILITY_MISMATCH)

            val expectedCredential: MobileCredentialOwner
            val expectedContinuity: MobileContinuityOwner
            val required: Set<MobileRouteCapability>
            val allowed: Set<MobileRouteCapability>
            val expectedTerminal: Boolean
            when (kind) {
                MobileItemKind.SAVED_CONNECTION -> {
                    expectedCredential = MobileCredentialOwner.SSH_CREDENTIAL
                    expectedContinuity = MobileContinuityOwner.REMOTE_TMUX_IF_ENABLED
                    required = setOf(MobileRouteCapability.TERMINAL_OUTPUT)
                    allowed = setOf(
                        MobileRouteCapability.TERMINAL_OUTPUT,
                        MobileRouteCapability.TERMINAL_INPUT,
                        MobileRouteCapability.TERMINAL_RESIZE,
                        MobileRouteCapability.PERSISTENT_TMUX,
                    )
                    expectedTerminal = true
                }
                MobileItemKind.PAIRED_DEVICE -> {
                    expectedCredential = MobileCredentialOwner.DEVICE_PAIRING_IDENTITY
                    expectedContinuity = MobileContinuityOwner.NONE
                    required = setOf(MobileRouteCapability.LIST_DEVICE_SESSIONS)
                    allowed = required
                    expectedTerminal = false
                }
                MobileItemKind.DURABLE_DEVICE_SESSION -> {
                    expectedCredential = MobileCredentialOwner.DEVICE_PAIRING_IDENTITY
                    expectedContinuity = MobileContinuityOwner.HOST_SERVICE
                    required = setOf(
                        MobileRouteCapability.TERMINAL_OUTPUT,
                        MobileRouteCapability.DURABLE_REPLAY,
                        MobileRouteCapability.AUTHORITATIVE_ACTIVITY,
                        MobileRouteCapability.SINGLE_WRITER,
                    )
                    allowed = setOf(
                        MobileRouteCapability.TERMINAL_OUTPUT,
                        MobileRouteCapability.TERMINAL_INPUT,
                        MobileRouteCapability.TERMINAL_RESIZE,
                        MobileRouteCapability.DURABLE_REPLAY,
                        MobileRouteCapability.AUTHORITATIVE_ACTIVITY,
                        MobileRouteCapability.SINGLE_WRITER,
                    )
                    expectedTerminal = true
                }
            }
            if (credential != expectedCredential) fail(MobileRouteContractError.CREDENTIAL_OWNER_MISMATCH)
            if (continuity != expectedContinuity) fail(MobileRouteContractError.CONTINUITY_OWNER_MISMATCH)
            if (!parsed.containsAll(required) || !allowed.containsAll(parsed)) {
                fail(MobileRouteContractError.CAPABILITY_MISMATCH)
            }
            if (canOpenTerminal != expectedTerminal) {
                fail(MobileRouteContractError.TERMINAL_OWNERSHIP_MISMATCH)
            }
            return MobileRouteProjection(kind, credential, continuity, parsed, canOpenTerminal)
        }

        private fun fail(error: MobileRouteContractError): Nothing = throw MobileRouteContractException(error)
    }
}

enum class MobileRootDestination { CONNECTIONS, DEVICES }

data class MobileTerminalDestination(
    val id: String,
    val title: String,
    val badge: String,
    val route: MobileRouteProjection,
) {
    init {
        if (!route.canOpenTerminal || route.itemKind == MobileItemKind.PAIRED_DEVICE) {
            throw MobileRouteContractException(MobileRouteContractError.TERMINAL_OWNERSHIP_MISMATCH)
        }
    }
}

File: apps/ios/TermiRustMobile/Models/MobileRouteContract.swift
import Foundation

enum MobileItemKind: String, Codable, Sendable {
    case savedConnection = "saved_connection"
    case pairedDevice = "paired_device"
    case durableDeviceSession = "durable_device_session"
}

enum MobileCredentialOwner: String, Codable, Sendable {
    case sshCredential = "ssh_credential"
    case devicePairingIdentity = "device_pairing_identity"
}

enum MobileContinuityOwner: String, Codable, Sendable {
    case none
    case remoteTmuxIfEnabled = "remote_tmux_if_enabled"
    case hostService = "host_service"
}

enum MobileRouteCapability: String, Codable, CaseIterable, Sendable {
    case listDeviceSessions = "list_device_sessions"
    case terminalOutput = "terminal_output"
    case terminalInput = "terminal_input"
    case terminalResize = "terminal_resize"
    case persistentTmux = "persistent_tmux"
    case durableReplay = "durable_replay"
    case authoritativeActivity = "authoritative_activity"
    case singleWriter = "single_writer"
}

enum MobileRouteContractError: String, Error, Equatable, Sendable {
    case unknownItemKind = "unknown_item_kind"
    case unknownCapability = "unknown_capability"
    case credentialOwnerMismatch = "credential_owner_mismatch"
    case continuityOwnerMismatch = "continuity_owner_mismatch"
    case capabilityMismatch = "capability_mismatch"
    case terminalOwnershipMismatch = "terminal_ownership_mismatch"
}

struct MobileRouteProjection: Equatable, Sendable {
    let itemKind: MobileItemKind
    let credentialOwner: MobileCredentialOwner
    let continuityOwner: MobileContinuityOwner
    let capabilities: Set<MobileRouteCapability>
    let canOpenTerminal: Bool

    static func validated(
        itemKind rawItemKind: String,
        credentialOwner rawCredentialOwner: String,
        continuityOwner rawContinuityOwner: String,
        capabilities rawCapabilities: [String],
        canOpenTerminal: Bool
    ) throws -> MobileRouteProjection {
        guard let itemKind = MobileItemKind(rawValue: rawItemKind) else {
            throw MobileRouteContractError.unknownItemKind
        }
        guard let credentialOwner = MobileCredentialOwner(rawValue: rawCredentialOwner) else {
            throw MobileRouteContractError.credentialOwnerMismatch
        }
        guard let continuityOwner = MobileContinuityOwner(rawValue: rawContinuityOwner) else {
            throw MobileRouteContractError.continuityOwnerMismatch
        }
        let parsedCapabilities = rawCapabilities.compactMap(MobileRouteCapability.init(rawValue:))
        guard parsedCapabilities.count == rawCapabilities.count else {
            throw MobileRouteContractError.unknownCapability
        }
        let capabilities = Set(parsedCapabilities)
        guard capabilities.count == rawCapabilities.count else {
            throw MobileRouteContractError.capabilityMismatch
        }

        let expectedCredential: MobileCredentialOwner
        let expectedContinuity: MobileContinuityOwner
        let required: Set<MobileRouteCapability>
        let allowed: Set<MobileRouteCapability>
        let expectedTerminal: Bool
        switch itemKind {
        case .savedConnection:
            expectedCredential = .sshCredential
            expectedContinuity = .remoteTmuxIfEnabled
            required = [.terminalOutput]
            allowed = [.terminalOutput, .terminalInput, .terminalResize, .persistentTmux]
            expectedTerminal = true
        case .pairedDevice:
            expectedCredential = .devicePairingIdentity
            expectedContinuity = .none
            required = [.listDeviceSessions]
            allowed = required
            expectedTerminal = false
        case .durableDeviceSession:
            expectedCredential = .devicePairingIdentity
            expectedContinuity = .hostService
            required = [.terminalOutput, .durableReplay, .authoritativeActivity, .singleWriter]
            allowed = [
                .terminalOutput, .terminalInput, .terminalResize, .durableReplay,
                .authoritativeActivity, .singleWriter,
            ]
            expectedTerminal = true
        }
        guard credentialOwner == expectedCredential else {
            throw MobileRouteContractError.credentialOwnerMismatch
        }
        guard continuityOwner == expectedContinuity else {
            throw MobileRouteContractError.continuityOwnerMismatch
        }
        guard required.isSubset(of: capabilities), capabilities.isSubset(of: allowed) else {
            throw MobileRouteContractError.capabilityMismatch
        }
        guard canOpenTerminal == expectedTerminal else {
            throw MobileRouteContractError.terminalOwnershipMismatch
        }
        return MobileRouteProjection(
            itemKind: itemKind,
            credentialOwner: credentialOwner,
            continuityOwner: continuityOwner,
            capabilities: capabilities,
            canOpenTerminal: canOpenTerminal
        )
    }
}

enum MobileRootDestination: String, CaseIterable, Sendable {
    case connections
    case devices
}

struct MobileTerminalDestination: Identifiable, Equatable, Sendable {
    let id: String
    let title: String
    let badge: String
    let route: MobileRouteProjection

    init(id: String, title: String, badge: String, route: MobileRouteProjection) throws {
        guard route.canOpenTerminal, route.itemKind != .pairedDevice else {
            throw MobileRouteContractError.terminalOwnershipMismatch
        }
        self.id = id
        self.title = title
        self.badge = badge
        self.route = route
    }
}
